src/PostgREST/RequestIntent.hs

```diff
@@ -0,0 +1,104 @@
+module PostgREST.RequestIntent where
+
+import qualified Data.Aeson           as JSON
+import qualified Data.ByteString      as BS
+import qualified Data.ByteString.Lazy as BL
+import           Data.List            (find)
+import           Data.Maybe           (fromMaybe, isJust, isNothing,
+                                       listToMaybe)
+import           Network.Wai          (Request (..))
+import           Network.Wai.Parse    (parseHttpAccept)
+import           PostgREST.RangeQuery (NonnegRange, rangeRequested)
+import           PostgREST.Types      (QualifiedIdentifier (..), Schema)
+
+type RequestBody = BL.ByteString
+
+-- | Types of things a user wants to do to tables/views/procs
+data Action = ActionCreate | ActionRead
+            | ActionUpdate | ActionDelete
+            | ActionInfo   | ActionInvoke
+-- | The target db object of a user action
+data Target = TargetIdent QualifiedIdentifier
+            | TargetRoot
+-- | Enumeration of currently supported content types for
+-- route responses and upload payloads
+data ContentType = ApplicationJSON | TextCSV
+-- | When Hasql supports the COPY command then we can
+-- have a special payload just for CSV, but until
+-- then CSV is converted to a JSON array
+data Payload = PayloadJSON JSON.Array
+
+-- | Describes what the user wants to do. This data type is a
+-- translation of the raw elements of an HTTP request into domain
+-- specific language.  There is no guarantee that the intent is
+-- sensible, it is up to a later stage of processing to determine
+-- if it is an action we are able to perform.
+data Intent = Intent {
+  -- | Set to Nothing for unknown HTTP verbs
+    iAction :: Maybe Action
+  -- | Set to Nothing for malformed range
+  , iRange  :: Maybe NonnegRange
+  -- | Set to Nothing for strangely nested urls
+  , iTarget :: Maybe Target
+  -- | The content type the client most desires (or JSON if undecided)
+  , iAccepts :: Either BS.ByteString ContentType
+  -- | {foo} becomes [{foo}] and CSV is converted to JSON
+  , iPayload :: Maybe Payload
+  -- | Taken from JSON Web Token
+  , iTrustedClaims :: Maybe JSON.Object
+  -- | If client wants created items echoed back
+  , iPreferRepresentation :: Bool
+  -- | If client wants first row as raw object
+  , iPreferSingular :: Bool
+  }
+
+-- | Examines HTTP request and translates it into user intent.
+userIntent :: Schema -> Request -> RequestBody -> Intent
+userIntent schema req _ =
+  let action = case requestMethod req of
```

Comment:
These let's contain more logic than the where's so by putting them in the function I guess I'm trying to call attention to their logic. Let's do kind of suck for indentation though and they require the ...`in` ending.